### PR TITLE
Enable Gradient Checkpointing for UNet2DModel (New)

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR supports gradient checkpointing for `UNet2DModel`.

Successor to PR #6718.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sayakpaul
@yiyixuxu

### CODE REVIEW

**Review by sayakpaul — APPROVED** [review]
Clean!

**Comment by yiyixuxu** [comment]
sorry @dg845 
don't know how I missed this PR! would you be able to resolve the conflicts and we'll merge once CI is green?

**Comment by dg845** [comment]
@yiyixuxu resolved the merge conflicts and updated the code to match the way gradient checkpointing is currently done in `UNet2DConditionModel`. The tests are passing locally for me. (Also, messed up the git history a little bit, whoops😅).

**Review by sayakpaul — APPROVED** [review]
Thanks a lot!

**Comment by yiyixuxu** [comment]
@hlky can you take a look here and help merge in?

File: src/diffusers/models/unets/unet_2d.py [hunk 1]
```diff
@@ -240,6 +242,10 @@ def __init__(
         self.conv_act = nn.SiLU()
         self.conv_out = nn.Conv2d(block_out_channels[0], out_channels, kernel_size=3, padding=1)
 
+    def _set_gradient_checkpointing(self, module, value=False):
+        if hasattr(module, "gradient_checkpointing"):
+            module.gradient_checkpointing = value
```
Comment (hlky) [inline src/diffusers/models/unets/unet_2d.py:247]: I notice we use `hasattr(module, "gradient_checkpointing")` in around half the cases of `_set_gradient_checkpointing` and module names in the other half, no need to change it for now but we could look at making it uniform at some point.

**Review by hlky — COMMENTED** [review]
Thanks! I've left a general comment and I'll look into the test failures, should be good to merge after CI turns green.

**Review by hlky — APPROVED** [review]
Thanks @dg845!
